// node-themoviedb/dist/utils/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class BaseError extends Error {
    constructor(code, message, errorCode) {
        super(message);
        this.code = code;
        this.errorCode = errorCode;
    }
}
class BadRequest extends BaseError {
    constructor() {
        super(0, 'Bad Request', 400);
    }
}
exports.BadRequest = BadRequest;
class UnauthorizedError extends BaseError {
    constructor(code, message) {
        super(code, message, 401);
    }
}
exports.UnauthorizedError = UnauthorizedError;
class NotFoundError extends BaseError {
    constructor(code, message) {
        super(code, message, 404);
    }
}
exports.NotFoundError = NotFoundError;
class RequestTimeout extends BaseError {
    constructor() {
        super(0, 'Request Timeout', 408);
    }
}
exports.RequestTimeout = RequestTimeout;
class TooManyRequests extends BaseError {
    constructor(retryAfter) {
        super(0, 'Too many requests per IP', 429);
        this.retryAfter = retryAfter;
    }
}
exports.TooManyRequests = TooManyRequests;
